modules/core/src/lifecycle/component-state.ts

```diff
@@ -19,74 +19,70 @@
 // THE SOFTWARE.
 
 import {isAsyncIterable} from '../utils/iterable-utils';
-import {PROP_SYMBOLS} from './constants';
-const {ASYNC_ORIGINAL, ASYNC_RESOLVED, ASYNC_DEFAULTS} = PROP_SYMBOLS;
+import {ASYNC_ORIGINAL, ASYNC_RESOLVED, ASYNC_DEFAULTS} from './constants';
+import type Component from './component';
+import {ComponentProps, StatefulComponentProps} from './component';
+import {PropType} from './prop-types';
 
 const EMPTY_PROPS = Object.freeze({});
 
-export default class ComponentState {
-  constructor(component = null) {
+type AsyncPropState = {
+  type: PropType;
+  lastValue: any;
+  resolvedValue: any;
+  pendingLoadCount: number;
+  resolvedLoadCount: number;
+};
+
+export default class ComponentState<T extends ComponentProps> {
+  component: Component<T>;
+  onAsyncPropUpdated: (propName: keyof T, value: any) => void;
+
+  private asyncProps: Partial<Record<keyof T, AsyncPropState>>;
```

Comment:
Nit: Not a super fan of such complicated/abstract typings. Keeping things simple to read is sometimes better. Though I suppose this is pretty low level.

Reply:
I actually find adding the typing help with the readability of this class. Not defining the type clearly would not change how this internal logic functions. If you find the type confusing maybe it's an indication that the implementation itself is problematic.